da/da.go

```diff
@@ -236,9 +238,174 @@ func (dac *DAClient) RetrieveHeaders(ctx context.Context, dataLayerHeight uint64
+			dac.logger.Info(message)
+			break
+		}
+		blobSize += uint64(len(blob))
+		blobs = append(blobs, blob)
+	}
+	if len(blobs) == 0 {
+		return ResultSubmitBatch{
+			BaseResult: BaseResult{
+				Code:    StatusError,
+				Message: "failed to submit blocks: no blobs generated " + message,
+			},
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, dac.SubmitTimeout)
+	defer cancel()
+	ids, err := dac.DA.Submit(ctx, blobs, gasPrice, dac.Namespace)
+	if err != nil {
+		status := StatusError
+		switch {
+		case errors.Is(err, &ErrTxTimedOut{}):
+			status = StatusNotIncludedInBlock
+		case errors.Is(err, &ErrTxAlreadyInMempool{}):
+			status = StatusAlreadyInMempool
+		case errors.Is(err, &ErrTxIncorrectAccountSequence{}):
+			status = StatusAlreadyInMempool
+		case errors.Is(err, &ErrTxTooLarge{}):
+			status = StatusTooBig
+		case errors.Is(err, &ErrContextDeadline{}):
+			status = StatusContextDeadline
+		}
+		return ResultSubmitBatch{
+			BaseResult: BaseResult{
+				Code:    status,
+				Message: "failed to submit block data: " + err.Error(),
+			},
+		}
+	}
+
+	if len(ids) == 0 {
+		return ResultSubmitBatch{
+			BaseResult: BaseResult{
+				Code:    StatusError,
+				Message: "failed to submit data: unexpected len(ids): 0",
+			},
+		}
+	}
+
+	return ResultSubmitBatch{
+		BaseResult: BaseResult{
+			Code:           StatusSuccess,
+			DAHeight:       binary.LittleEndian.Uint64(ids[0]),
+			BlobSize:       blobSize,
+			SubmittedCount: uint64(len(ids)),
+		},
+	}
+}
+
+// RetrieveBatch retrieves block data from DA.
+func (dac *DAClient) RetrieveBatch(ctx context.Context, dataLayerHeight uint64) ResultRetrieveBatch {
+	idsResult, err := dac.DA.GetIDs(ctx, dataLayerHeight, dac.Namespace)
+	if err != nil {
+		return ResultRetrieveBatch{
+			BaseResult: BaseResult{
+				Code:     StatusError,
+				Message:  fmt.Sprintf("failed to get IDs: %s", err.Error()),
+				DAHeight: dataLayerHeight,
+			},
+		}
+	}
+	ids := idsResult.IDs
+
+	// If no block data are found, return a non-blocking error.
+	if len(ids) == 0 {
+		return ResultRetrieveBatch{
+			BaseResult: BaseResult{
+				Code:     StatusNotFound,
+				Message:  (&ErrBlobNotFound{}).Error(),
+				DAHeight: dataLayerHeight,
+			},
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, dac.RetrieveTimeout)
+	defer cancel()
+	blobs, err := dac.DA.Get(ctx, ids, dac.Namespace)
+	if err != nil {
+		return ResultRetrieveBatch{
+			BaseResult: BaseResult{
+				Code:     StatusError,
+				Message:  fmt.Sprintf("failed to get blobs: %s", err.Error()),
+				DAHeight: dataLayerHeight,
+			},
+		}
+	}
+
+	data := make([]*types.Data, len(blobs))
+	for i, blob := range blobs {
+		var d pb.Data
+		err = proto.Unmarshal(blob, &d)
+		if err != nil {
+			dac.logger.Error("failed to unmarshal block data", "daHeight", dataLayerHeight, "position", i, "error", err)
+			continue
+		}
+		data[i] = new(types.Data)
+		err := data[i].FromProto(&d)
+		if err != nil {
+			return ResultRetrieveBatch{
```

Comment:
Few lines before, there's a `continue` on error, but here error is returned immediately - why?

Reply:
this was the implementation within centralize sequencer. I can dig into it but im not sure myself. Maybe @gupadhyaya knows?

Reply:
junk data is something that cannot be unmarshalled to types.Data, hence we ignore it. if it is a valid data, it better be able to `.FromProto` it, so returning error.

Reply:
@tzdybal i think we had some discussion on this long time back. Again, this is copied from rollkit.